x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,847 @@
+            }
+            return deleteResult;
+        }
+
+        @Override
+        public void deleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException {
+            Objects.requireNonNull(blobNames);
+
+            // find all the blob names that must be deleted
+            Set<String> blobNamesSet = new HashSet<>(blobNames);
+            Set<String> blobNamesToDelete = new HashSet<>();
+            for (String existingBlobName : delegatedBlobContainer.listBlobs().keySet()) {
+                if (blobNamesSet.contains(existingBlobName)) {
+                    blobNamesToDelete.add(existingBlobName);
+                }
+            }
+
+            // find all the metadata blob names that must be deleted
+            Map<String, List<String>> blobNamesToMetadataNamesToDelete = new HashMap<>(blobNamesToDelete.size());
+            Set<String> allMetadataBlobNames = new HashSet<>();
+            try {
+                allMetadataBlobNames = encryptionMetadataBlobContainer.listBlobs().keySet();
+            } catch (IOException e) {
+                // the metadata blob container might not even exist
+                // the encrypted data is the "anchor" for encrypted blobs, if those are removed, the encrypted blob as a whole is
+                // considered removed, even if, technically, the metadata is still lingering (it should later be removed by cleanup)
+                // therefore this tolerates metadata delete failures, when data deletes are successful
+                logger.warn("Failure to list blobs of metadata blob container " + encryptionMetadataBlobContainer.path(), e);
+            }
+            for (String metadataBlobName : allMetadataBlobNames) {
+                final String blobNameForMetadata;
+                try {
+                    blobNameForMetadata = MetadataIdentifier.parseFromMetadataBlobName(metadataBlobName).v1();
+                } catch (IllegalArgumentException e) {
+                    // ignore invalid metadata blob names, which most likely have been created externally
+                    continue;
+                }
+                // group metadata blob names to their associated blob name
+                if (blobNamesToDelete.contains(blobNameForMetadata)) {
+                    blobNamesToMetadataNamesToDelete.computeIfAbsent(blobNameForMetadata, k -> new ArrayList<>(1))
+                            .add(metadataBlobName);
+                }
+            }
+            // Metadata deletes when there are multiple for the same blob is un-safe, so don't try it now.
+            // It is unsafe because metadata "appears" before the data and there could be an overwrite in progress for which only
+            // the metadata, but not the encrypted data, shows up.
+            List<String> metadataBlobNamesToDelete = new ArrayList<>(blobNamesToMetadataNamesToDelete.size());
```

Comment:
This doesn't matter IMO. On e.g. S3 we have no ordering guarantees at all anyway so we should just delete everything here. There is no point in trying to protect ourselves against concurrency issues at this layer. If the caller wants a certain blob name gone, the fact that someone may be writing this very same blob name right now is never an issue due to the repository consistency measures we have in place for blob naming.

Reply:
The code here is mostly to protect for the write after delete cases.
I understand this is not a concern because of the code higher up, but I would still have this protection.
I don't think discounting that case buys us anything wrt to efficiency with cloud API calls.